### ISSUE TITLE
Non symmetric block

### ISSUE THREAD

**Issue by bserranoanton** [OSUPCVLab/SegFormer3D#6]
Hi all, 

I want to implement a solution with input shape (B, 1, 256, 256, 32), so blocks are not symmetric. I expect the output shape to be the same as the input shape instead for the number of channels (classes): (B, 7, 256, 256, 32). Any idea how to implement this?

Thanks

**Comment by srperera** [comment]
Hi yes, this is possible. Just to clarify is (B, 1, 256, 256, 32) = (Batch Size, Features, H, W, D) or you want to follow the pytorch 3D tensor configuration where (Batch Size, Features, Depth, H, W) because it seems like the 256x256 might be the spatial resolution.

**Comment by bserranoanton** [comment]
Hi, yes, data is in the format (Batch Size, Features, H, W, D)  but it is not a problem to modify the dataloader to have data in format (Batch Size, Features, Depth, H, W). 2D images correspond to a different slices of a 3D volume.

Many thanks.

**Comment by srperera** [comment]
So this is going to be a quite long reply just so I address and explain the details so I can make it as easy as possible for you to get up and running. All of these are explained in the file im attaching too. You can CTRL-F "#MODIFYHERE" to find the two places in the code you would want to modify if other configurations are needed however it might take you a little while to figure out the exact dimensions. Parameters and explanations for 3 possible variants are provided below that kind of gives you a trade off between full resolution vs efficiency. Please let me know if you have any questions. 

### #################################################################################
      # Custom Parameteres For [B, C, 32, 256, 256]
      """ 
      The trick that worked for us is modifying the Depth Kernel/Strid/Padding
      such that within the first or second blocks we reduce the D, H, W size
      disparity. This unsymmetric [D, H, W] dims causes some shape issues 
      that we need to pay attention to.
      Example for the parameters below for a tensor of shape [B, C, 32, 256, 256]
      the tensor sizes at each block is as follows
          * Block 1 -- [B, C, 32, 64, 64]
          * Block 2 -- [B, C, 32, 32, 32]
          * Block 2 -- [B, C, 16, 16, 16]
          * Block 2 -- [B, C, 8, 8, 8]
      For data of shape [B, C, 32, 256, 256] you might want to play with the 
      "sr_ratios" parameter because you might benefit from a larger ratio 
      for the first Block since the squence length from my initial check
      is quite large. 
      
      Additionally if you dont want to modify the squence length but you want 
      to improve the computational complexity or fit larger batch sizes you can
      resize the data to [B, C, 32, 128, 128] or [B, C, 16, 128, 128]. 
      
      For [B, C, 32, 128, 128] (Spatial Only Resize Keeping Num Frames In Tact)
      Possible parameters V1 which would result in following shapes at each block:
          * Block 1 -- [B, C, 32, 32, 32]
          * Block 2 -- [B, C, 16, 16, 16]
          * Block 2 -- [B, C, 8, 8, 8]
          * Block 2 -- [B, C, 4, 4, 4]
          patch_kernel_size = [
                  [3, 7, 7],  # MODIFIED Originally (7, 7, 7)
                  [3, 3, 3],
                  [3, 3, 3],
                  [3, 3, 3],
              ]

          patch_stride = [
                  [1, 4, 4],  # MODIFIED Originally (4, 4, 4)
                  [2, 2, 2],
                  [2, 2, 2],
                  [2, 2, 2],
              ]

          patch_padding = [
                  [1, 3, 3],  # MODIFIED Originally (3, 3, 3)
                  [1, 1, 1],
                  [1, 1, 1],
                  [1, 1, 1],
              ]
      
      For [B, C, 16, 128, 128] (Spatial Only Resize Keeping Num Frames In Tact)
      The possible parameters are (same as for [B, C, 32, 256, 256]) because 
      it will just halve the spatial dimension.
          * Block 1 -- [B, C, 16, 32, 32]
          * Block 2 -- [B, C, 16, 16, 16]
          * Block 2 -- [B, C, 8, 8, 8]
          * Block 2 -- [B, C, 4, 4, 4]
          patch_kernel_size = [
              [3, 7, 7],  # MODIFIED Originally (7, 7, 7)
              [3, 3, 3],
              [3, 3, 3],
              [3, 3, 3],
          ]

          patch_stride = [
              [1, 4, 4],  # MODIFIED Originally (4, 4, 4)
              [1, 2, 2],
              [2, 2, 2],
              [2, 2, 2],
          ]

          patch_padding = [
              [1, 3, 3],  # MODIFIED Originally (3, 3, 3)
              [1, 1, 1],
              [1, 1, 1],
              [1, 1, 1],
          ]
          
      Note the at the original architecture does the follow size reduction
      so a input resolution of [B, C, 256, 256, 256] ie symmetric input would be 
          * Block 1 - (D/4, H/4, W/4) -> output shape = [64, 64, 64]
          * Block 2 - (D/8, H/8, W/8) -> output shape = [32, 32, 32]
          * Block 3 - (D/16, H/16, W/16) -> output shape = [16, 16, 16]
          * Block 4 - (D/32, H/32, W/32) -> output shape = [8, 8, 8]
      The parameters I provided above sections generates a similar progression
      in the way the spatial resolutions are modified within the architecture.  
      """
####################################################################################

Hope this is clear. Please see the architecture file too. its already written for the dimensions you specified. 
[segformer3d_nonsymmetric.txt](https://github.com/user-attachments/files/15982363/segformer3d_nonsymmetric.txt)

**Comment by bserranoanton** [comment]
Many thanks for the explanation and the code :) 
When running the code I'm using input shape size=(1, 1, 32, 256, 256). In my case:
in_channels = 1
num_classes = 7
I'm expecting the same number of channels in the output but I get [1, 7, 128, 256, 256] (128 instead of 32). How can I fix that? 

Many thanks again

**Comment by srperera** [comment]
Hi sorry about that I think I had some typo in the original file. Here are all three files again. I ran each of them and tested the inputs and outputs to match your experiment settings. 

Updated Files:
[segformer3d_nonsymmetric.txt](https://github.com/user-attachments/files/16032961/segformer3d_nonsymmetric.txt)
[segformer3d_nonsymmetric_v2.txt](https://github.com/user-attachments/files/16032962/segformer3d_nonsymmetric_v2.txt)
[segformer3d_nonsymmetric_v3.txt](https://github.com/user-attachments/files/16032963/segformer3d_nonsymmetric_v3.txt)

Tested Sample Input/Ouputs:
![non_symmetric_output](https://github.com/OSUPCVLab/SegFormer3D/assets/44714882/da41c8b6-6139-4701-8712-dec7359ff81e)

**Comment by bserranoanton** [comment]
Hi, 

Yes, now shapes are correct! Many, many thanks :)
